pkg/asset/manifests/dns.go

```diff
@@ -134,14 +146,37 @@ func (d *DNS) Generate(dependencies asset.Parents) error {
 			}
 		}
 	case gcptypes.Name:
-		if installConfig.Config.Publish == types.ExternalPublishingStrategy {
-			zone, err := icgcp.GetPublicZone(context.TODO(), installConfig.Config.Platform.GCP.ProjectID, installConfig.Config.BaseDomain)
+
+		// Set the public zone
+		switch {
+		case installConfig.Config.Publish != types.ExternalPublishingStrategy:
+			// Do not use a public zone when not publishing externally.
+		case installConfig.Config.GCP.PublicDNSZone.ID != "":
+			// Use the provided zone if specified.
+			project := findGCPProject(installConfig.Config.GCP.PublicDNSZone.ProjectID, installConfig.Config.GCP.ProjectID)
+			config.Spec.PublicZone = &configv1.DNSZone{ID: combineGCPZoneInfo(project, installConfig.Config.GCP.PublicDNSZone.ID)}
+		default:
+			// Search the project for a zone with the specified base domain.
+			project := findGCPProject(installConfig.Config.GCP.PublicDNSZone.ProjectID, installConfig.Config.GCP.ProjectID)
+			zone, err := icgcp.GetPublicZone(context.TODO(), project, installConfig.Config.BaseDomain)
 			if err != nil {
 				return errors.Wrapf(err, "failed to get public zone for %q", installConfig.Config.BaseDomain)
 			}
-			config.Spec.PublicZone = &configv1.DNSZone{ID: zone.Name}
+			config.Spec.PublicZone = &configv1.DNSZone{ID: combineGCPZoneInfo(project, zone.Name)}
```

Comment:
This is the same breaking change as above... More particular, here we know the project has not been overrided.